### PR TITLE
Create UnknownPropertiesTest.java

### PR DESCRIPTION
**Pull request by pjfanning**
test case for https://github.com/FasterXML/jackson-databind/issues/3721

### CODE REVIEW

**Comment by cowtowncoder** [comment]
Ok so this exhibits the problem: and it reminds me that it is specifically the `@JsonCreator` that induces the issue here -- and I think with Scala (module) essentially all POJOs (POSOs?) use that handling.

Aside from that, I guess testing/verification here would mean adding breakpoints (etc) in accessor or such -- or just observing time to run?

**Comment by pjfanning** [comment]
Add a breakpoint in NumberInput.parseBigInteger (both versions of this overloaded method).
Scala norms are to use `case classes` for JavaBean equivalent use cases but case classes are immutable (by default) - ExtractFieldsNoDefaultConstructor is a Java approximation of how case classes work.

**Comment by cowtowncoder** [comment]
@pjfanning Ah-ha! Ok, yes, now I see it clearly: `BeanDeserializer`, line 503 or so:

```
            if (_ignoreAllUnknown) {
                p.skipChildren();
                continue;
            }
```

but for some reason not otherwise. I think I'll see how many tests break if using this for all cases...

**Comment by cowtowncoder** [comment]
Ohhhhh. Actually, the test failing in `TestPolymorphicCreators` is quite interesting: it is NOT using `@JsonTypeInfo` based polymorphism, but rather manual kind wherein `@JsonCreator` in super type is used but then when instance is constructed, we actually look for potentially different deserializer  (see `handlePolymorphic` of `BeanDeserializer`).
That is quite a powerful feature and no doubt used by some developers somewhere.

And all that is required is the use of properties-based creator.

But this also suggests "ignoreUnknown" would break this usage too. Weird.

**Comment by cowtowncoder** [comment]
Ok, quick note: I may not have explained my ideas clearly, but the lazy/buffered value types must be provided by `jackson-core`, not create by databind. This because while getting text via `JsonParser.getText()` works for JSON backend, it is likely going to be pretty bad for binary formats.

But maybe work here could try prototyping approach with easier-to-test approach. Even if eventual implementation would be different and split across components.

**Comment by pjfanning** [comment]
> Ok, quick note: I may not have explained my ideas clearly, but the lazy/buffered value types must be provided by `jackson-core`, not create by databind. This because while getting text via `JsonParser.getText()` works for JSON backend, it is likely going to be pretty bad for binary formats.
> 
> But maybe work here could try prototyping approach with easier-to-test approach. Even if eventual implementation would be different and split across components.

How about if I move the new Lazy* classes to jackson-core and add getLazyBigIntegerValue() and getLazyBigDecimalValue() to JsonParser interface?

**Comment by cowtowncoder** [comment]
> > Ok, quick note: I may not have explained my ideas clearly, but the lazy/buffered value types must be provided by `jackson-core`, not create by databind. This because while getting text via `JsonParser.getText()` works for JSON backend, it is likely going to be pretty bad for binary formats.
> > But maybe work here could try prototyping approach with easier-to-test approach. Even if eventual implementation would be different and split across components.
> 
> How about if I move the new Lazy* classes to jackson-core and add getLazyBigIntegerValue() and getLazyDecimalValue() to JsonParser interface?

That would be better, although I am not sure there should be separate accessors for types, instead of something like `getLazyNumber()` -- given that caller (TokenBuffer) does not necessarily have same information on actual underlying type (binary formats have strict types; JSON has approximate type, configuration dependant).
It all has to do with call sequence used for buffering.

File: src/main/java/com/fasterxml/jackson/databind/util/TokenBuffer.java [hunk 1]
```diff
@@ -442,6 +449,8 @@ public void serialize(JsonGenerator gen) throws IOException
                         gen.writeNumber((Long) n);
                     } else if (n instanceof Short) {
                         gen.writeNumber((Short) n);
+                    } else if (n instanceof LazyNumber) {
+                        gen.writeNumber(((LazyNumber) n).getText());
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/databind/util/TokenBuffer.java:453]: Potential problem here would be that binary formats do not support "Number as Strings".
But there is method `JsonGenerator.canWriteFormattedNumbers()` that actually allows checking if generator implements support for `writeNumber(String)` or not (word "formatted" is bit misnomer here).

Comment (pjfanning) [reply src/main/java/com/fasterxml/jackson/databind/util/TokenBuffer.java]: I have now added a new local method that checks `canWriteFormattedNumbers`

**Comment by pjfanning** [comment]
replaced with #3761

**Pull request closed without merging**